Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx

```diff
@@ -0,0 +1,361 @@
+
+    ic.services().get<CallbackService>().set<CallbackService::Id::Stop>([this]() {
+      LOG(info) << "Saving records into ROOT file";
+      mAlign.terminate();
+    });
+  }
+
+  //_________________________________________________________________________________________________
+  void finaliseCCDB(framework::ConcreteDataMatcher& matcher, void* obj)
+  {
+    /// finalize the track extrapolation setting
+    if (mCCDBRequest && base::GRPGeomHelper::instance().finaliseCCDB(matcher, obj)) {
+      if (matcher == framework::ConcreteDataMatcher("GLO", "GRPMAGFIELD", 0)) {
+        TrackExtrap::setField();
+        TrackExtrap::useExtrapV2();
+        mAlign.SetBFieldOn(mch::TrackExtrap::isFieldON());
+        trackFitter.smoothTracks(true);
+      }
+
+      if (matcher == framework::ConcreteDataMatcher("GLO", "GEOMALIGN", 0)) {
+        LOG(info) << "Loading reference geometry from CCDB";
+        transformation = geo::transformationFromTGeoManager(*gGeoManager);
+        for (int i = 0; i < 156; i++) {
+          int iDEN = GetDetElemId(i);
+          transform[iDEN] = transformation(iDEN);
+        }
+      }
+    }
+  }
+
+  //_________________________________________________________________________________________________
+  void run(framework::ProcessingContext& pc)
+  {
+    if (mCCDBRequest) {
+      base::GRPGeomHelper::instance().checkUpdates(pc);
+    }
+
+    o2::globaltracking::RecoContainer recoData;
+    recoData.collectData(pc, *mDataRequest.get());
+
+    const auto& mchTracks = recoData.getMCHTracks();
+    const auto& mchmidMatches = recoData.getMCHMIDMatches();
+    const auto& mchClusters = recoData.getMCHTrackClusters();
+
+    for (auto const& mchmidMatch : mchmidMatches) {
+
+      int mchTrackID = mchmidMatch.getMCHRef().getIndex();
+      const auto& mchTrack = mchTracks[mchTrackID];
+      int first = mchTrack.getFirstClusterIdx();
+      int last = mchTrack.getLastClusterIdx();
+      int clIndex = -1;
+      mch::Track convertedTrack;
+
+      for (int i = first; i <= last; i++) {
+        clIndex += 1;
+        const auto& cluster = mchClusters[i];
+        mch::Cluster* mch_cluster = new mch::Cluster();
+        mch_cluster->x = cluster.x;
+        mch_cluster->y = cluster.y;
+        mch_cluster->z = cluster.z;
+
+        uint32_t ClUId = mch::Cluster::buildUniqueId(int(cluster.getDEId() / 100) - 1, cluster.getDEId(), clIndex);
+        mch_cluster->uid = ClUId;
+
+        mch_cluster->ex = cluster.ex;
+        mch_cluster->ey = cluster.ey;
+
+        convertedTrack.createParamAtCluster(*mch_cluster);
+      }
+
+      if (convertedTrack.getNClusters() > 9) {
+        // Erase removable track
+        if (!RemoveTrack(convertedTrack)) {
+          // mRecords.emplace_back(mAlign.ProcessTrack(convertedTrack, transformation, false, weightRecord));
+          pc.outputs().snapshot(Output{"MUON", "RECORD_MCHMID", 0}, mAlign.ProcessTrack(convertedTrack, transformation, false, weightRecord));
```

Comment:
not sure if/how this work, but in principle this should be done once per TF (i.e. per `run(...)`), probably sending one vector of `MillePedeRecord` in this case.

Reply:
This actually works for my case; I tested with a vector valued branch like what you have done for MCH tracks and clusters. But for the case of alignment, it seems to me more convenient to save directly a TTree of records so MillePede2 can directly read and use it. For the above one, from the test I've done with 3 CTFs, I actually didn't capture any trouble.
I'm not sure if I understood correctly: by one vector you mean I keep the current definition of branch but instead of sending records one by one I save them in a vector when doing loop then send the resulting vector directly to output?

Reply:
I am also not sure how this will work, perhaps @ktf can comment if creating multiple snapshots of the same output in a single TF will not lead to only last one or 1st one actually received by the consumer. But for sure, if in the TF you don't produce an output the DPL will warn you about the lost TF, since you define the `{"MUON", "RECORD_MCHMID", 0}` with `Lifetime::Timeframe`. If the output is sporadic, you better define https://github.com/AliceO2Group/AliceO2/blob/5d3fdf914b7208018a78f6500d9325ee21c06caf/Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx#L346 with `Lifetime::Sporadic`.